### PR TITLE
[SPARK-25154][SQL] Support NOT IN sub-queries inside nested OR conditions

### PR DESCRIPTION
**Pull request by maropu**
### What changes were proposed in this pull request?

Currently NOT IN subqueries (predicated null aware subquery) are not allowed inside OR expressions. We currently catch this condition in checkAnalysis and throw an error.

This PR enhances the subquery rewrite to support this type of queries.

Query
```SQL
SELECT * FROM s1 WHERE a > 5 or b NOT IN (SELECT c FROM s2);
```
Optimized Plan
```SQL
== Optimized Logical Plan ==
Project [a#3, b#4]
+- Filter ((a#3 > 5) || NOT exists#7)
   +- Join ExistenceJoin(exists#7), ((b#4 = c#5) || isnull((b#4 = c#5)))
      :- HiveTableRelation `default`.`s1`, org.apache.hadoop.hive.serde2.lazy.LazySimpleSerDe, [a#3, b#4]
      +- Project [c#5]
         +- HiveTableRelation `default`.`s2`, org.apache.hadoop.hive.serde2.lazy.LazySimpleSerDe, [c#5, d#6]
```
This is rework from #22141.
The original author of this PR is @dilipbiswal.

Closes #22141

### Why are the changes needed?

For better usability.

### Does this PR introduce any user-facing change?

No.

### How was this patch tested?

Added new tests in SQLQueryTestSuite, RewriteSubquerySuite and SubquerySuite.
Output from DB2 as a reference:
[nested-not-db2.txt](https://github.com/apache/spark/files/2299945/nested-not-db2.txt)

### CODE REVIEW

**Comment by maropu** [comment]
retest this please

File: sql/core/src/test/resources/sql-tests/results/subquery/in-subquery/nested-not-in.sql.out [hunk 1]
```diff
@@ -0,0 +1,332 @@
...
+       dept_id 
+FROM   emp 
+WHERE  id = 600
+       OR id = 500 
+       OR dept_id NOT IN (SELECT dept_id 
+                          FROM   emp)
+-- !query schema
+struct<id:int,dept_id:int>
+-- !query output
+500	NULL
```
Comment (maropu) [inline sql/core/src/test/resources/sql-tests/results/subquery/in-subquery/nested-not-in.sql.out:97]: I've checked that the output is the same with the PostgreSQL answers.

**Comment by maropu** [comment]
cc: @cloud-fan @dongjoon-hyun

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/subquery.scala [hunk 2]
```diff
@@ -162,12 +162,21 @@ object RewritePredicateSubquery extends Rule[LogicalPlan] with PredicateHelper {
...
           val exists = AttributeReference("exists", BooleanType, nullable = false)()
           newPlan =
             buildJoin(newPlan, sub, ExistenceJoin(exists), conditions.reduceLeftOption(And))
           exists
+        case Not(InSubquery(values, ListQuery(sub, conditions, _, _))) =>
+          val exists = AttributeReference("exists", BooleanType, nullable = false)()
+          // Deduplicate conflicting attributes if any.
+          val newSub = dedupSubqueryOnSelfJoin(newPlan, sub, Some(values))
+          val inConditions = values.zip(sub.output).map(EqualTo.tupled)
+          val nullAwareJoinConds = inConditions.map(c => Or(c, IsNull(c)))
```
Comment (cloud-fan) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/subquery.scala:176]: can we add some comments to explain it?

File: sql/core/src/test/resources/sql-tests/results/subquery/in-subquery/nested-not-in.sql.out (hunk 1 above)
Comment (cloud-fan) [reply sql/core/src/test/resources/sql-tests/results/subquery/in-subquery/nested-not-in.sql.out:97]: does our ported pgsql test include "nested not in" queries?

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/subquery.scala (hunk 2 above)
Comment (maropu) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/subquery.scala:176]: ok

File: sql/core/src/test/resources/sql-tests/results/subquery/in-subquery/nested-not-in.sql.out (hunk 1 above)
Comment (maropu) [reply sql/core/src/test/resources/sql-tests/results/subquery/in-subquery/nested-not-in.sql.out:97]: Ah, yea, we need to check that. I'll do later.

Comment (maropu) [reply sql/core/src/test/resources/sql-tests/results/subquery/in-subquery/nested-not-in.sql.out:97]: I checked the ported tests don't have those queries. Actually, the original regression tests in PostgreSQL don't have, too.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/subquery.scala (hunk 2 above)
Comment (maropu) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/subquery.scala:187]: Added.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/subquery.scala [hunk 3]
```diff
@@ -162,12 +162,32 @@ object RewritePredicateSubquery extends Rule[LogicalPlan] with PredicateHelper {
...
           val exists = AttributeReference("exists", BooleanType, nullable = false)()
           newPlan =
             buildJoin(newPlan, sub, ExistenceJoin(exists), conditions.reduceLeftOption(And))
           exists
+        case Not(InSubquery(values, ListQuery(sub, conditions, _, _))) =>
+          val exists = AttributeReference("exists", BooleanType, nullable = false)()
+          // Deduplicate conflicting attributes if any.
+          val newSub = dedupSubqueryOnSelfJoin(newPlan, sub, Some(values))
+          val inConditions = values.zip(sub.output).map(EqualTo.tupled)
+          // To handle null-aware predicate sub-queries in nested conditions,
```
Comment (cloud-fan) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/subquery.scala:176]: where do we define "null-aware predicate sub-queries"? It's only `InSubquery`?

Comment (maropu) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/subquery.scala:176]: Yea, IIUC `null-aware predicate sub-queries` just means `InSubquery` judging from the `hasNullAwarePredicateWithinNot` implementation. https://github.com/apache/spark/pull/28158/files#diff-34d3859f7c9fd82a4b286686c21d26c2L112

**Comment by maropu** [comment]
retest this please

File: sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/optimizer/RewriteSubquerySuite.scala [hunk 4]
```diff
@@ -52,4 +52,21 @@ class RewriteSubquerySuite extends PlanTest {
...
   }
 
+  test("NOT-IN subquery nested inside OR") {
+    val relation1 = LocalRelation('a.int, 'b.int)
+    val relation2 = LocalRelation('c.int, 'd.int)
+    val exists = 'exists.boolean.notNull
+
+    val query = relation1.where('b === 1 || Not('a.in(ListQuery(relation2.select('c))))).select('a)
+
+    val plan = relation1.select('a).where('b === 1 || Not('exists))
```
Comment (cloud-fan) [inline sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/optimizer/RewriteSubquerySuite.scala:62]: where do we use `plan`?

**Review by cloud-fan — APPROVED** [review]

**Comment by maropu** [comment]
Thanks, @cloud-fan @dilipbiswal ! Merged to master.

**Pull request closed without merging**